### PR TITLE
feat: add store and store service for categories

### PR DESCRIPTION
**Pull request by Nadiia-Valkiv-NerdySoft**
## Links

https://trello.com/c/wFp0Cnav/9-implement-quizzes-catalog-page

## Problem

Not store at project

## Solution

Add elf store for saving categories

### CODE REVIEW

File: src/app/core/services/categoriesStore.service.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import { Injectable } from '@angular/core';
```
Comment (IhorChai) [inline src/app/core/services/categoriesStore.service.ts:1]: Let's keep the file names consistent throughout the project (kebab-case)

File: src/app/pages/catalog/catalog.component.ts [hunk 2]
```diff
@@ -14,9 +15,25 @@ import { AsyncPipe } from '@angular/common';
...
+
   categories$!: Observable<QuizCategory[]>;
 
   ngOnInit(): void {
-    this.categories$ = this.categoriesService.getRandomCategories();
+    this.categories$ = this.quizCategoriesStoreService.getCategories();
+
+    this.categories$.subscribe((categories) => {
+      if (!categories || categories.length === 0) {
+        this.loadCategories();
```
Comment (IhorChai) [inline src/app/pages/catalog/catalog.component.ts:29]: Components should not know about store manipulation methods. They should only query the data.
Store addition and update should be processed in the API component. You can add there an RxJs tap() operator and update the store from there

File: src/app/core/services/categoriesStore.service.ts
Comment (IhorChai) [inline src/app/core/services/categoriesStore.service.ts:1]: It's also better to split the service into 2 - one for mutations and one for querying

**Review by OlhaNerdySoft — APPROVED** [review]

**Review by IhorChai — APPROVED** [review]

File: src/app/pages/catalog/catalog.component.ts [hunk 3]
```diff
@@ -14,9 +14,18 @@ import { AsyncPipe } from '@angular/common';
 })
 export class CatalogComponent implements OnInit {
   private readonly categoriesService = inject(CategoriesService);
-  categories$!: Observable<QuizCategory[]>;
+  private readonly categoriesStoreService = inject(CategoriesStoreService);
+
+  categories$!: Observable<any>;
 
   ngOnInit(): void {
-    this.categories$ = this.categoriesService.getRandomCategories();
```
Comment (IhorChai) [inline src/app/pages/catalog/catalog.component.ts:20]: how to  optimise this code:

  ngOnInit(): void {
    this.categoriesService.getRandomCategories().subscribe();
     
    this.categories$ = this.categoriesStoreService.getCategories()
}
    
In categories service:

getRandomCategories(): Observable<QuizCategory[]> { 
    if(categories && categories.length > 0) {
        return this.categoriesStoreService.getCategories()
    }
    
    return this.apiService.fetchCategories().pipe(
      map(categories => this.randomizationService
      .getRandomItems(categories, this.categoriesNumber)
      .map(category => this.enrichCategory(category))),
            tap((categories) => {
        this.categoriesStoreService.addCategories(categories);
      }),
    );
}

File: src/app/core/services/categories.service.ts [hunk 4]
```diff
@@ -1,39 +1,37 @@
...
-    return this.http
-    .get<{ trivia_categories: QuizCategory[] }>(this.categoriesUrl)
-    .pipe(
-      map(response => response.trivia_categories),
-      catchError((error) => {
-        console.error('Error fetching categories:', error);
-        return of([]);
+      tap((categories) => {
+        this.categoriesStoreService.addCategories(categories);
+        this.categories = categories;
```
Comment (IhorChai) [inline src/app/core/services/categories.service.ts:34]: No need to save categories in service you can access static data from the store or use one of their caching methods
https://ngneat.github.io/elf/docs/features/requests/requests-cache/

**Pull request merged**